shared/actions/fs.js

```diff
@@ -20,6 +21,150 @@ import {
 import {isWindows} from '../constants/platform'
 import {saveAttachmentDialog, showShareActionSheet} from './platform-specific'
 import {type TypedState} from '../util/container'
+import {FolderTypeToString} from '../constants/rpc'
+import {findKey} from 'lodash'
+
+// Take the parsed JSON from kbfs/favorite/list, and populate an array of
+// Types.FolderRPCWithMeta with the appropriate metadata:
+//
+// 1) Is this favorite ignored or new?
+// 2) Does it need a rekey?
+//
+const _fillMetadataInFavoritesResult = (
+  favoritesResult: Object,
+  myKID: any
+): Array<Types.FolderRPCWithMeta> => {
+  const fillFolder = toMerge => folder => {
+    folder.waitingForParticipantUnlock = []
+    folder.youCanUnlock = []
+    Object.keys(toMerge).forEach(key => {
+      folder[key] = toMerge[key]
+    })
+
+    if (!folder.problem_set) {
+      return
+    }
+
+    const solutions = folder.problem_set.solution_kids || {}
+    if (Object.keys(solutions).length) {
+      folder.needsRekey = true
+    }
+
+    if (folder.problem_set.can_self_help) {
+      const mySolutions = solutions[myKID] || []
+      folder.youCanUnlock = mySolutions.map(kid => {
+        const device = favoritesResult.devices[kid]
+        return {...device, deviceID: kid}
+      })
+    } else {
+      folder.waitingForParticipantUnlock = Object.keys(solutions).map(userID => {
+        const devices = solutions[userID].map(kid => favoritesResult.devices[kid].name)
+        const numDevices = devices.length
+        const last = numDevices > 1 ? devices.pop() : null
+
+        return {
+          name: favoritesResult.users[userID],
+          devices: `Tell them to turn on${numDevices > 1 ? ':' : ' '} ${devices.join(', ')}${
+            last ? ` or ${last}` : ''
+          }.`,
+        }
+      })
+    }
+  }
+
+  favoritesResult.favorites.forEach(fillFolder({isIgnored: false, isNew: false}))
```

Comment:
I find this flow a little confusing. It might be simpler just to merge it yourself
```
  return [
      ...favoritesResult.favorites.map(f => {...fillFolder(f), isIgnored: false, isNew: false})),
      ...favoritesResult.ignored.map(f => {...fillFolder(f), isIgnored: true, isNew: false})),
      ...favoritesResult.new.map(f => {...fillFolder(f), isIgnored: false, isNew: true})),
  ]
```

Reply:
it also might help to move this kind of stuff to constants. Actions tends to have a lot of logic in it and these converters are pretty isolatable